### PR TITLE
Allow extensions to optionally add own description (on load)

### PR DESCRIPTION
**Pull request by carlopi**
This allows extensions to opt-in and provide their own description, to be later fetched from example like (after this PR):
```
D FROM duckdb_extensions() WHERE extension_name == 'shell';
┌────────────────┬─────────┬───────────┬──────────────┬───────────────────────────────────────────────┬───────────┬───────────────────┬───────────────────┬────────────────┐
│ extension_name │ loaded  │ installed │ install_path │                  description                  │  aliases  │ extension_version │   install_mode    │ installed_from │
│    varchar     │ boolean │  boolean  │   varchar    │                    varchar                    │ varchar[] │      varchar      │      varchar      │    varchar     │
├────────────────┼─────────┼───────────┼──────────────┼───────────────────────────────────────────────┼───────────┼───────────────────┼───────────────────┼────────────────┤
│ shell          │ true    │ true      │              │ Adds CLI-specific support and functionalities │ []        │                   │ STATICALLY_LINKED │                │
└────────────────┴─────────┴───────────┴──────────────┴───────────────────────────────────────────────┴───────────┴───────────────────┴───────────────────┴────────────────┘
```
versus current behaviour of:
```
D FROM duckdb_extensions() WHERE extension_name == 'shell';
┌────────────────┬─────────┬───────────┬──────────────┬─────────────┬───────────┬───────────────────┬───────────────────┬────────────────┐
│ extension_name │ loaded  │ installed │ install_path │ description │  aliases  │ extension_version │   install_mode    │ installed_from │
│    varchar     │ boolean │  boolean  │   varchar    │   varchar   │ varchar[] │      varchar      │      varchar      │    varchar     │
├────────────────┼─────────┼───────────┼──────────────┼─────────────┼───────────┼───────────────────┼───────────────────┼────────────────┤
│ shell          │ true    │ true      │              │             │ []        │                   │ STATICALLY_LINKED │                │
└────────────────┴─────────┴───────────┴──────────────┴─────────────┴───────────┴───────────────────┴───────────────────┴────────────────┘
```

Information is NOT propagated to ExtensionInstalInfo struct / file, so this is visible only for loaded extensions.

Allowing extensions to register their descriptions provide already some functionality and exposed an additional API to extension.
I would propose to review (as a follow up), whether it should be considered to persist to the extension_name.duckdb_extension.info file, but that can build on top of the capability provided here.

### CODE REVIEW

**Comment by carlopi** [comment]
@isaacbrodsky had a similar idea / PR, so pinging for feedback.

File: src/include/duckdb/main/database.hpp [hunk 1]
```diff
@@ -79,6 +84,7 @@ class DatabaseInstance : public enable_shared_from_this<DatabaseInstance> {
 	unique_ptr<ConnectionManager> connection_manager;
 	unordered_set<string> loaded_extensions;
 	unordered_map<string, ExtensionInstallInfo> loaded_extensions_data;
+	unordered_map<string, ExtensionLoadedInfo> loaded_extensions_extra_data;
```
Comment (Mytherin) [inline src/include/duckdb/main/database.hpp:87]: Looks like we now have three maps/sets that all have the exact same keys, can we merge `loaded_extensions`, `loaded_extensions_data` and `loaded_extensions_extra_data`?

Comment (carlopi) [reply src/include/duckdb/main/database.hpp:87]: I think that makes sense, sure.
One thing is that currently it's not a given that the 3 maps have the same cardinality / but having a map string to a struct like:
```
{
    bool isLoaded;
    unique_ptr<ExtensionInstallInfo> install_info;
    unique_ptr<ExtensionLoadedInfo> loaded_info;
}
```
plus having iterators check for validity is cleaner.

**Review by samansmink — APPROVED** [review]
LGTM! ideally we would have this information available before loading, but i guess that would mean storing it in the footer, which seems undesirable

**Comment by carlopi** [comment]
Thank you both for the feedback.

---

I was thinking that for the before-load (and for unifying what we do for our extensions) the path could be have an extension (statically loaded for core ones, loadable for community) provide the descriptions of other extensions.

That would be doable already with this PR, where we unify this code:
```c++
	auto alias_count = ExtensionHelper::ExtensionAliasCount();
	for (idx_t i = 0; i < extension_count; i++) {
		auto extension = ExtensionHelper::GetDefaultExtension(i);
		ExtensionInformation info;
		info.name = extension.name;
                 // ...
		info.description = extension.description;
         }
```
to a bunch of calls to:
```c++
     RegisterExtension(db, "extension_a", "My description");
```
performed elsewhere.

But my idea was to reconsider this later / first iterate solving the actual problem that is closing the gap between extensions and core extensions.

**Comment by Mytherin** [comment]
Could we not store it in the `.info` if having it available without loading is desirable?

**Comment by carlopi** [comment]
There are a few scenarios that might be relevant:

1. Extension is LOADed
2. Extension is INSTALLed, and has been loaded once
3. Extension is INSTALLed, never loaded
4. Extension is not installed, but known to be existent

This PR offers a way to solve 1, and infra to do 4, plus introduces the syntax to declare the descriptions in the code.
2 is basically very hard, since we don't want to write description in the header, but can be considered later.
3 is doable, if we append to the `.info` file on the first Load, but also this is to be considered later.

Checked again with @samansmink, I think this can be merged.

**Review by isaacbrodsky — APPROVED** [review]

**Pull request merged**